### DISCUSSION TITLE
How to write and publish reusable steps?

### DISCUSSION THREAD

**Discussion by ms-ati** [cucumber-rs/cucumber#236] [Q&A] [answered]
Hi! I am working on a port of [cucumber/aruba](https://github.com/cucumber/aruba) to `cucumber-rs`, to be called [aruba-rs](https://github.com/ms-ati/aruba-rs).

For anyone not familiar: Aruba, written in Ruby, is a supported part of Cucumber which provides a super useful set of Step Definitions and tools for testing **command-line applications**. I'd like to have these step definitions available when working on command-line applications in Rust in a way that is well-integrated and designed. Why? I've observed:
- A lot command-line applications being written in Rust
- A lack of common executable specification practices, leading to avoidable failure reports
- Similar patterns for using `cucumber-rs` to write specs for command-line apps, with unique issues and unforeseen edgecases, across many projects.

Wouldn't it be awesome if `aruba-rs` would lead to low-friction, widespread reuse of a shared set of step definitions for the executable specifications of command-line apps across the Rust ecosystem? 🙏🏻 🎈 

However, to get there, I have some questions! 😄 

**(1) How can a `dev-dependency` library publish reusable Step Definitions that require state?**

The challenge: As I understand it (please correct if I've misunderstood!), the design of `cucumber-rs` is for the step definition macros `given`/`when`/`then` to require, as their first parameter, a concrete struct for which we have already derived `cucumber::World` via a derive macro.

Q: How should a library such as `aruba-rs` make reusable step definitions available, where each step would need to know the ultimate concrete type of the `World` parameter, and access a common sub-set of state from it?

Assumption: I'll need to use the `World` parameter to access all mutable state about processes we have launched in the previous steps of any Scenario, for example:

```gherkin
Feature: Help text

  Scenario: If we pass the `--help` flag, we see usage documentation
    When I run `aruba_hello_world --help`
    Then the exit status should be 0
     And the stdout contains: "Usage: "
```

```rust
trait ArubaWorld {
    fn last_command_run(&self) -> &Option<std::process::Command>
}
```

**(2) How can `.feature` files in a project reference Step Definitions from a `dev-dependency` library?**

Thanks for any ideas!

**Comment by ilslv** [comment, marked as answer]
Very interesting use-case, thanks for bringing that up. I've created a separate repository as an example: https://github.com/ilslv/cucumber-steps-as-library 

> How can a dev-dependency library publish reusable Step Definitions that require state?

You should create a library, that depends on `cucumber` and exports `World` implementor with some step definitions. In example repo this is the `/steps` directory. In case you need additional steps, downstream crates can add them to an existing `World` implementor from another crate (`/tests/cucumber.rs` has an example).

The main limitation is that `World` implementor can't have generics, so I would advice adding something like [`anymap`] to it, so users could dynamically extend `World`. 




[`anymap`]: https://docs.rs/anymap/

**Comment by maximetinu** [comment]
Hi! I have found that this doesn't work if the opt-level is > 0. I've opened an issue about that: https://github.com/cucumber-rs/cucumber/issues/347
